File: python/three_sum_closest.py
from typing import List


class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        nums.sort()
        best_distance = float("inf")
        best_sum = 0
        n = len(nums)

        for i in range(0, n - 2):
            if i > 0 and nums[i] == nums[i - 1]:
                continue

            left = i + 1
            right = n - 1
            while left < right:
                current_sum = nums[right] + nums[left] + nums[i]

                if current_sum == target:
                    return current_sum
                elif current_sum > target:
                    right -= 1
                else:
                    left += 1

                current_distance = abs(current_sum - target)
                if current_distance < best_distance:
                    best_distance = current_distance
                    best_sum = current_sum

        return best_sum


s = Solution()

nums = [-1, 2, 1, -4]
target = 1
print(s.threeSumClosest(nums, target), 2)

nums = [0, 0, 0]
target = 1
print(s.threeSumClosest(nums, target), 0)

File: c++/three_sum_closest.cpp
#include <algorithm>
#include <iostream>
#include <vector>
using namespace std;

class Solution {
   public:
   public:
    int threeSumClosest(vector<int>& nums, int target) {
        sort(nums.begin(), nums.end());
        int bestSum = 0;
        int bestDistance = INT_MAX;
        int n = nums.size();

        for (int i = 0; i < n; i++) {
            if (i > 0 && nums[i] == nums[i - 1]) {
                continue;
            }

            int left = i + 1;
            int right = n - 1;
            while (left < right) {
                int currentSum = nums[i] + nums[left] + nums[right];
                if (currentSum == target) {
                    return currentSum;
                } else if (currentSum < target) {
                    left++;
                } else {
                    right--;
                }

                int currentDistance = abs(currentSum - target);
                if (currentDistance < bestDistance) {
                    bestDistance = currentDistance;
                    bestSum = currentSum;
                }
            }
        }

        return bestSum;
    }
};

int main() {
    Solution s;

    vector<int> nums1 = {-1, 2, 1, -4};
    int target1 = 1;
    cout << s.threeSumClosest(nums1, target1) << " Expected: 2" << endl;

    vector<int> nums2 = {0, 0, 0};
    int target2 = 1;
    cout << s.threeSumClosest(nums2, target2) << " Expected: 0" << endl;

    return 0;
}
